File: workspace/GSI_mutation++/src/main/GSI_mutation++.cpp
#include "input.h"
//#include <catch.hpp>
#include <Eigen/Dense>
#include <stdlib.h>
#include <stdio.h>

using namespace Mutation;
using namespace Eigen;
using namespace Mutation::Thermodynamics;
using namespace Mutation::Utilities;

int main(int argc, char *argv[])
{
    const double tol = std::numeric_limits<double>::epsilon();

    // Mixture
    //MixtureOptions opts("seb_oxidation_NASA9_ChemNonEq1T");

    //Read input
    Input input("input.xml");

    Mutation::GlobalOptions::workingDirectory(input.workingDirectory());
    Mutation::GlobalOptions::dataDirectory(input.dataDirectory());
    MixtureOptions opts(input.mixturename());
    Mixture mix(opts);
    const double P_init = input.pressureinit();
    const double T_init = input.temperatureinit();
    // Setting number of iterations for the solver
    mix.setIterationsSurfaceBalance(input.maxiterationsinit());
    mix.setSubIterationsSurfaceBalance(input.subiterationsinit());
    mix.setIterationsPert_m(input.pertminit());
    mix.setIterationsPert_T(input.pertTinit());
    mix.setIterationsEps(input.tolinit());
    bool iNewtonhistory;
    std::istringstream(input.iNewtonhistory()) >> std::boolalpha >> iNewtonhistory;
    mix.setIterationsHistory(iNewtonhistory);
    // Setting up
    const size_t set_state_with_rhoi_T = 1;
    const size_t pos_T_trans = 0;
    size_t ns = mix.nSpecies();
    size_t nT = mix.nEnergyEqns();
    size_t neq = ns + nT;
    // Conditions T = 3000K and p = 100Pa
    VectorXd Teq = VectorXd::Constant(nT, T_init);

    mix.equilibrate(Teq(pos_T_trans), P_init);

    // Mass gradient
    VectorXd xi_e(ns);
    VectorXd xi_ey(ns);
    const double dx = input.distanceinit();
    xi_ey = Map<const VectorXd>(mix.Y(), ns);
    xi_e = Map<const VectorXd>(mix.X(), ns);
    // std::cout <<xi_e<<xi_ey;
    //mix.getSpeciesComposition(std::string("Gas"), xi_e.data(),Composition::MOLE);
    mix.setDiffusionModel(xi_ey.data(), dx);

    // Temperature gradient
    VectorXd T_e = Teq;
    if (mix.getGSIMechanism() == "phenomenological_mass_energy")
        mix.setGasFourierHeatFluxModel(T_e.data(), dx);
        // mix.setGasFourierHeatFluxModel( 6000000.);

    // Initial conditions of the surface are the ones in the first
    // physical cell
    VectorXd rhoi_s(ns);
    mix.densities(rhoi_s.data());
    VectorXd T_s = VectorXd::Constant(nT, (T_init ));
    mix.setSurfaceState(rhoi_s.data(), T_s.data(), set_state_with_rhoi_T);

    // Solve balance and request solution
    mix.solveSurfaceBalance();
    mix.getSurfaceState(rhoi_s.data(), T_s.data(), set_state_with_rhoi_T);
    double rho = rhoi_s.sum();

    // Verifying the solution gives low residual in the balance equations
    mix.setState(rhoi_s.data(), T_s.data(), set_state_with_rhoi_T);
    VectorXd xi_s(ns);
    // xi_s = Map<const VectorXd>(mix.X(), ns);
    // Compute diffusion velocities
    
    VectorXd vdi(ns);
    // VectorXd dxidx(ns);
    // dxidx = (xi_s - xi_e) / dx;
    // double E = 0.;
    // mix.stefanMaxwell(dxidx.data(), vdi.data(), E);
    // std::cout <<vdi;
    // mix.setDiffusionModel(xi_ey.data(), dx);
    xi_s =rhoi_s/rhoi_s.sum();
    mix.comSurfaceDiffusionVelocity(xi_s.data(),vdi.data());

    // Conductive heat flux
    VectorXd gas_conduction(nT);
    if (mix.getGSIMechanism() == "phenomenological_mass_energy")
    {
        
        gas_conduction(pos_T_trans) = mix.computeGasFourierHeatFlux(T_s.data());
        // VectorXd dTdx(nT);
        // VectorXd lambda(nT);
        // dTdx = (T_s - T_e) / dx;
        // mix.frozenThermalConductivityVector(lambda.data());
        // std::cout<<mix.computeGasFourierHeatFlux(T_s.data());
    }

    // Get surface production rates
    VectorXd wdot(ns);
    mix.setSurfaceState(rhoi_s.data(), T_s.data(), set_state_with_rhoi_T);
    mix.surfaceReactionRates(wdot.data());

    // Blowing flux (should be zero for catalysis)
    double mblow;
    mix.getMassBlowingRate(mblow);

    // Species and mixture enthalpies
    VectorXd v_hi(ns * nT);
    mix.getEnthalpiesMass(v_hi.data());
    VectorXd v_h(nT);
    v_h(pos_T_trans) = mix.mixtureHMass();//(rhoi_s / rho).dot(v_hi.head(ns));

    // Surface radiation
    VectorXd q_srad = VectorXd::Zero(nT);
    if (mix.getGSIMechanism() == "phenomenological_mass_energy")
        q_srad(pos_T_trans) = mix.getSurfaceRadiativeHeatFlux();

    // Chemical Energy Contribution
    VectorXd v_hi_rhoi_vi = VectorXd::Zero(nT);
    v_hi_rhoi_vi(pos_T_trans) = v_hi.head(ns).dot(rhoi_s.cwiseProduct(vdi));

    //solid conduction
    VectorXd solid_conduction(nT);
    solid_conduction(pos_T_trans) = mix.computeSolidHeat();

    // Building balance functions
    VectorXd F(ns);
    if (mix.getGSIMechanism() == "phenomenological_mass_energy")
        F.resize(neq);
    F.head(ns) = (rhoi_s / rho) * mblow + rhoi_s.cwiseProduct(vdi) - wdot;
    if (mix.getGSIMechanism() == "phenomenological_mass_energy")
        F.tail(nT) = gas_conduction - q_srad + mblow * v_h +
                     v_hi_rhoi_vi + solid_conduction;

    // Compute error and write
    //double err = F.lpNorm<Infinity>();
    std::cout << std::endl;
    std::cout << "The residual of gas surface interaction:" << std::endl;
    std::cout << std::setw(18) << "Res.AVG(mass)";
    std::cout << std::setw(18) << "Res.MAX(mass)";
    std::cout << std::setw(18) << "Res.AVG(energy)";
    std::cout << std::setw(18) << "Res.MAX(energy)";
    std::cout << std::endl;
    std::cout << std::setw(18) << F.head(ns).lpNorm<1>();
    std::cout << std::setw(18) << F.head(ns).lpNorm<Infinity>();
    std::cout << std::setw(18) << F.tail(nT).lpNorm<1>();
    std::cout << std::setw(18) << F.tail(nT).lpNorm<Infinity>();
    std::cout << std::endl;
    // Write a header line for the table
    //Temperature
    std::cout << std::setw(28) << "Surface balance temperature[K]" << std::endl;
    for (int j = 0; j < nT; ++j)
        std::cout << std::setw(28) << T_s(j);
    std::cout << std::endl;

    std::cout << std::setw(28) << "Surface blowing flux[kg/m^2-s]" << std::endl;
    std::cout << std::setw(28) << mblow;
    std::cout << std::endl;

    std::cout << "Surface mass properties" << std::endl;
    std::cout << setw(30) << "";
    for (int j = 0; j < ns; ++j)
        std::cout << std::setw(16) << mix.speciesName(j);
    std::cout << std::endl;

    //species density
    std::cout << setw(30) << "Species density[kg/m^3]";
    for (int j = 0; j < ns; ++j)
        std::cout << std::setw(16) << rhoi_s(j);
    std::cout << std::endl;
    //chemical source
    std::cout << setw(30) << "Chemical production[kg/m^2-s]";
    for (int j = 0; j < ns; ++j)
        std::cout << std::setw(16) << wdot(j);
    std::cout << std::endl;
    std::cout << setw(30) << "Surface species mole fraction";
    for (int j = 0; j < mix.nSpecies(); ++j)
        std::cout << std::setw(16) << mix.X()[j];
    std::cout << std::endl;
    std::cout << setw(30) << "Edge species mole fraction";
    for (int j = 0; j < mix.nSpecies(); ++j)
        std::cout << std::setw(16) << xi_e[j];
    std::cout << std::endl;

    cout.precision(4);
    //设定后续以科学计数法的方式输出浮点数
    cout.setf(ios::scientific);

    std::cout << "Surface energy properties" << std::endl;
    std::cout << std::setw(22) << "Gas conductive heat[J]";
    std::cout << std::setw(18) << "Radiative heat[J]";
    std::cout << std::setw(18) << "Blowing heat[J]";
    std::cout << std::setw(18) << "Diffusion heat[J]";
    std::cout << std::setw(26) << "Solid conductive heat[J]";
    std::cout << std::endl;
    for (int j = 0; j < nT; ++j)
    {
        std::cout << std::setw(22) << (gas_conduction(j));
        std::cout << std::setw(18) << (-q_srad(j));
        std::cout << std::setw(18) << (mblow * v_h(j));
        std::cout << std::setw(18) << (-v_hi_rhoi_vi(j));
        std::cout << std::setw(26) << (mblow * solid_conduction(j));
        std::cout << std::endl;
    }


}

File: workspace/GSI_mutation++/src/main/GSI_mutation++.f90
program main
    use mutationpp
    use input
    implicit none
    integer(kind=4) :: i, j, ne, ns, nt, neq
    integer (IP) :: set_state_with_rhoi_T, pos_T_trans

    real(kind=8) :: T, P, rho, E, mblow, solid_heat
    character(len=12), allocatable, DIMENSION(:) :: species_name
    real(RP), dimension(:), allocatable :: rhoi_s, wdot, xi_s, xi_e, mwi, dxidx, vdi, v_hi, v_h
    real(RP), dimension(:), allocatable :: T_e, T_s, dtdx, lambda, q_srad, q_gcon, v_hi_rhoi_vi, Residual

    set_state_with_rhoi_T = 1
    pos_T_trans = 1

    call setInput

    call MPP_globaloptionsworkingdir(m_working_directory)
    call mpp_initializegsi(m_mixture_name)

    !set solver options
    call mpp_setiterationssurfacebalance(m_maxstep)
    call mpp_setiterationspert_m(m_pert_m)
    call mpp_setiterationspert_t(m_pert_T)
    call mpp_setiterationseps(m_tol)
    call mpp_setiterationshistory(m_iNewtonhistory)
    call mpp_get_gsi_mechism(m_gsi_mechism)

    ns = mpp_nspecies()
    nt = mpp_n_energy_eqns()
    ne = mpp_nelements()
    neq = ns + nt

    allocate (xi_e(ns), rhoi_s(ns), xi_s(ns), dxidx(ns), vdi(ns), wdot(ns), Residual(ns+1))
    allocate (T_e(nt), T_s(nt))
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        allocate (dtdx(nt), lambda(nt), q_srad(nt), q_gcon(nt))
    end if
    allocate (v_hi(ns*nt), v_h(nt), v_hi_rhoi_vi(nt))
    allocate (species_name(ns))

    !obtain species name list
    do j = 1, ns
        call mpp_species_name(j, species_name(j))
    end do

    !set diffusion model use equilibrium composition of the mixture with given T,P
    call mpp_equilibrate(m_temperature_init, m_pressure_init)
    call mpp_x(xi_e)
    !boundary edge composition can be given by input
    !call mpp_get_species_composition("Gas",xi_e)
    call mpp_set_diffusion_model(xi_e, m_distance)

    !set heat gradient if needed
    T_e = m_temperature_init
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        call mpp_set_cond_heat_flux(T_e, m_distance)
    end if

    !Initial conditions of the surface are the ones in the first physical cell
    call mpp_species_densities(rhoi_s)
    T_s = m_temperature_init/2.0
    call mpp_set_surface_state(rhoi_s, T_s, set_state_with_rhoi_T)

    ! Solve balance and request solution
    call mpp_solve_surface_balance()
    call mpp_get_surface_state(rhoi_s, T_s, set_state_with_rhoi_T)
    rho = sum(rhoi_s)

    ! Verifying the solution gives low residual in the balance equations
    call mpp_set_state(rhoi_s, T_s, set_state_with_rhoi_T)
    call mpp_x(xi_s)

    ! compute diffusion velocities
    dxidx = (xi_s - xi_e)/m_distance; 
    E = 0.0
    call mpp_stefan_maxwell(dxidx, vdi, E)

    ! compute heat flux
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        !dtdx = (T_s - T_e)/m_distance
        !call mpp_frozen_thermal_conductivity(lambda)
        !q_gcon = lambda*dtdx
        !gas heat flux can be calculated by (-lambda*dtdx=q_gcon>0) or function below, just for validation
        call mpp_compute_gas_heat_flux(T_s,q_gcon)
    end if

    ! Get surface production rates
    call mpp_set_surface_state(rhoi_s, T_s, set_state_with_rhoi_T)
    call mpp_surface_production_rates(wdot)

    ! Blowing flux (should be zero for catalysis)
    call mpp_mass_blowing_rate(mblow)

    ! Species and mixture enthalpies
    call mpp_species_h_mass(v_hi)
    v_h(pos_T_trans) = DOT_PRODUCT(rhoi_s(1:ns)/rho, v_hi(1:ns))

    ! Surface radiation
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        q_srad = 0.0
        q_srad(pos_T_trans) = mpp_getsurfaceradiativeheatflux()
    end if

    ! Chemical Energy Contribution
    v_hi_rhoi_vi(pos_T_trans) = dot_product((v_hi(1:ns)), (rhoi_s*vdi))

    ! solid conduction
    solid_heat=mblow*mpp_get_solid_heat()
    ! MixtureOptions graphiteopt("graphite.xml");
    ! Mixture graphite(graphiteopt);
    ! const int set_state_PT = 1;
    ! graphite.setState(&P_init,&T_s[0],1);
    ! double hcp=graphite.mixtureHMass(T_s[0]);

    ! Building balance functions
    Residual(1:ns) = (rhoi_s/rho)*mblow + (rhoi_s*vdi) - wdot; 
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        Residual(ns + 1) = sum(q_gcon) - q_srad(pos_T_trans) + mblow*v_h(pos_T_trans) + v_hi_rhoi_vi(pos_T_trans) + solid_heat
    end if

    ! write results
    write (*, *) "The residual of gas surface interaction:"
    write (*, "(4A18)") "Res.AVG(mass)", "Res.MAX(mass)", "Res.AVG(energy)", "Res.MAX(energy)"
    write (*, "(4E18.7)") sum(Residual(1:ns))/ns, maxval(Residual(1:ns)), sum(Residual(ns + 1:ns+1)), maxval(Residual(ns + 1:ns+1))

    write (*, *) "Surface balance temperature[K]"
    do i = 1, nt
        write (*, '(f18.5)', advance='no') T_s(i)
    end do
    write (*, *)

    write (*, *) "Surface blowing flux[kg/m^2-s]"
    write (*, "(f18.7)") mblow

    write (*, "(A30)", advance='no') "Surface mass properties"
    do i = 1, ns
        write (*, '(5X,A13)', advance='no') species_name(i)
    end do
    write (*, *)

    ! species density
    write (*, "(A30)", advance='no') "Species density[kg/m^3]"
    do i = 1, ns
        write (*, "(E18.6)", advance='no') rhoi_s(i)
    end do
    write (*, *)

    ! chemical source
    write (*, "(A30)", advance='no') "Chemical production[kg/m^2-s]"
    do i = 1, ns
        write (*, "(E18.6)", advance='no') wdot(i)
    end do
    write (*, *)

    write (*, "(A30)", advance='no') "Surface species mole fraction"
    do i = 1, ns
        write (*, "(E18.6)", advance='no') xi_s(i)
    end do
    write (*, *)

    write (*, "(A30)", advance='no') "Edge species mole fraction"
    do i = 1, ns
        write (*, "(E18.6)", advance='no') xi_e(i)
    end do
    write (*, *)

    write (*, *) "Surface energy properties"
    write (*, "(5A25)") "Conductive heat[J]", "Radiative heat[J]", "Blowing heat[J]", "Diffusion heat[J]", "Solid heat(J)"
    write (*, "(5E25.7)", advance='no') sum(q_gcon), -q_srad(pos_T_trans) , &
    mblow*v_h(pos_T_trans), v_hi_rhoi_vi(pos_T_trans), solid_heat
    write (*, *)

    deallocate (xi_e, rhoi_s, xi_s, dxidx, vdi, wdot, Residual)
    deallocate (T_e, T_s)
    if (m_gsi_mechism == "phenomenological_mass_energy") then
        deallocate (dtdx, lambda, q_srad, q_gcon)
    end if
    deallocate (v_hi, v_h, v_hi_rhoi_vi)
    deallocate (species_name)
    call mpp_destroy()
101 format(1X, F18.7)

end program
